department-of-veterans-affairs/caseflow
# Prevent Virtual Hearing modal from being submitted multiple times

Resolves #14201

### Description

- Add disabled state to link-style button
- Add validation to `VirtualHearing` modal to disallow creating a `VirtualHearing` for a virtual hearing
- Add validation to `base_hearing_update_form` to disallow creating or updating a virtual hearing that has an active job running
  - Add new error message if this validation step fails
- Refactors in the `VirtualHearingModal`
  - Fix some issues with how the email form is pre-populated
  - Also fixes #13947
- Disable buttons and email inputs on `VirtualHearingModal` when waiting for response from server
  - Add loading state to submit button
- Add and update tests around the modal and update form

### Acceptance Criteria
- [x] The submit button on the VirtualHearingModal enters the Loading state after the first click
- [x] Verify that the form cannot be submitted multiple times during a single interaction
- [x] Protect on the back end against multiple active virtual hearings being created for the same hearing
- [x] Tests (don't bother writing a feature test for multiple submissions, but please provide steps to test manually in the pull request)

### Testing Plan
1. Comment out line that starts the `CreateConferenceJob` in `base_hearing_update_form.rb`. Commenting this out simulates a scenario where the conference is being updated, but the job still hasn't run. 
2. Go to Hearing Details for a Video Hearing as a Hearing Coordinator
3. Copy the URL, and paste it in another tab
4. Go to the first tab, and switch the hearing to virtual
5. In the second tab, try to perform the same action
6. The modal should close, and you should see an error message

### User Facing Changes

#### Failure Message

<img width="1251" alt="Screen Shot 2020-05-11 at 5 40 25 PM" src="https://user-images.githubusercontent.com/1298274/81614805-8e58d380-93ae-11ea-9e2e-13e957fa8ff3.png">

---

<img width="1251" alt="Screen Shot 2020-05-11 at 5 41 42 PM" src="https://user-images.githubusercontent.com/1298274/81614873-af212900-93ae-11ea-90e1-135e60492675.png">

#### Loading State

<img width="736" alt="Screen Shot 2020-05-11 at 4 06 59 PM" src="https://user-images.githubusercontent.com/1298274/81606646-7464c400-93a1-11ea-9887-22c36b9129b1.png">

---

<img width="736" alt="Screen Shot 2020-05-11 at 4 04 14 PM" src="https://user-images.githubusercontent.com/1298274/81606440-19cb6800-93a1-11ea-96c3-f6ccf2d2e710.png">
